### DISCUSSION TITLE
How to build MicroPython with multiple external user C modules

### DISCUSSION THREAD

**Discussion by saraverbeecke** [micropython/micropython#11660] [Libraries & Drivers] [answered]
Hello,

I want to build MicroPython with 2 external user C modules namely: st7789 and ulab.
I can build MicroPython with one of them, but I don't know how to add multiple C user modules in the make command... 
Example for an ESP32:
```shell
make BOARD=GENERIC_S3_SPIRAM_OCT USER_C_MODULES="$REPO_ROOT/ulab/code/micropython.cmake"
``` 

Thanks,
Sara

**Comment by Wind-stormger** [comment]
I care about this too.

**Comment by cwalther** [comment]
This seems well documented at https://docs.micropython.org/en/latest/develop/cmodules.html#compiling-the-cmodule-into-micropython, is there anything wrong with it? I haven’t done this myself on CMake-based ports so I cannot comment.

**Comment by ricksorensen** [comment, marked as answer]
I am doing this currently.  I have a common `.cmake` file that invokes the `cmake` files for the modules I wish to build ... for example:

In `mymods` directory I have the command cmake file and the (git cloned) st7789 and ulab source:
```
$ls -1 mymods -F
micropython.cmake
st7789_mpy/
ulab/
```
and the cmake file is:
```
$ cat mymods/micropython.cmake 
include(${CMAKE_CURRENT_LIST_DIR}/ulab/code/micropython.cmake)
include(${CMAKE_CURRENT_LIST_DIR}/st7789_mpy/st7789/micropython.cmake)
```
I can then build with
```
make USER_C_MODULES="${full_path}/mymods/micropython.cmake"
```

**Reply by saraverbeecke** [reply]
Thanks that is what I needed.

**Comment by trentp-igor** [comment]
Could `USER_C_MODULES` be a list?  Seems like that way it wouldn't be necessary to add additional files to build more than one extension.

**Comment by andrewleech** [comment]
The standard usage is to have `USER_C_MODULES` pointing at a folder which contains one or more c extensions to include. 

The example in  https://docs.micropython.org/en/latest/develop/cmodules.html#compiling-the-cmodule-into-micropython is set to https://github.com/micropython/micropython/tree/master/examples/usercmodule which contains three separate extensions.

On `make` based ports this is all you need to do, make finds and includes all modules in that folder.

It's only for the more annoying `cmake` based ports (like esp) that you need to manually list the modules you want, as per https://github.com/micropython/micropython/blob/master/examples/usercmodule/micropython.cmake

**Comment by xyzzy42** [comment]
It seems like most external modules have their code nested some level under the root.  For example, `ulab/code`.  So if someone wants to pull in multiple extensions as git submodules, they're going to have paths in the project root like:
```
modules/ulab/code
modules/st7789_mpy/st7789
```
To put them all a single directory would require another directory of relative symlinks.

If USER_C_MODULES was a list, then it would be possible to inject multiple external modules with only a change to build parameters and it would work if the modules can't all be in a single directory because the upstream projects that provide them are not organized this way.
